Docker image for `optimade` on ghcr.io (#1171)
Materials-Consortia/optimade-python-tools: INSTALL.md

```

The following command starts a local MongoDB instance, runs the test suite, then stops and deletes the containers:

```shell
docker run -d --name mongo_test -p 27017:27017 -d mongo:4.4.6 \
&& OPTIMADE_DATABASE_BACKEND="mongodb" py.test; \
docker container stop mongo_test; docker container rm mongo_test
```

## Container image

### Retrieve the image

The [`optimade` container image](https://github.com/Materials-Consortia/optimade-python-tools/pkgs/container/optimade) is available from the [GitHub Container registry](https://ghcr.io).

> To pull the latest version using Docker run the following:

Reviewer: ```suggestion
To pull the latest version using [Docker](https://docs.docker.com/) run the following:
```

```shell
docker pull ghcr.io/materials-consortia/optimade:latest
```

!!! note

>     The `:latest` part can be left out, as the `latest` version will be pulled by default.

Reviewer: ```suggestion
    The tag, `:latest`, can be left out, as the `latest` version will be pulled by default.
```

If you'd like to pull a specific version, this can be done by replacing `latest` in the command above with the version of choice, e.g., `0.17.1`.
To see which versions are available, please go [here](https://github.com/Materials-Consortia/optimade-python-tools/pkgs/container/optimade/versions).

### Run a container

When starting a container from the image there are a few choices.

> It is possible to run both the standard OPTIMADE server implemented in this repository as well as an index meta-database from this image.

Reviewer: ```suggestion
It is possible to run either a standard OPTIMADE server, or an [index meta-database](https://github.com/Materials-Consortia/OPTIMADE/blob/master/optimade.rst#index-meta-database) server from this image.
```

Author: I just added the extra sentence here to underline my original point.

The key is setting the environment variable `MAIN`.

| **MAIN** | **Result** |
|:---:|:--- |
| `main` | Standard OPTIMADE server. |
| `main_index` | Index meta-database OPTIMADE server. |

Using Docker, the following command will run a container from the image:

```shell

> # rm will remove container when it exits.

Reviewer 2: Nice that you briefly explain what each flag does :+1:

# detach will run the server in the background.
# publish will run the server from the host port 8080.
# name will give the container a handy name for referencing later.
docker run \
    --rm \
    --detach \
    --publish 8080:5000 \
    --env MAIN=main \
    --name my-optimade \
    ghcr.io/materials-consortia/optimade:latest
```

The server should now be available at [localhost:8080](http://localhost:8080).

### Configure the server

> The server will use test data and a test configuration by default.

Reviewer: ```suggestion
By default, the server will use the test configuration that includes test data for structures and references and an in-memory mongomock database backend.
```

Author: Changed the wording slightly, since it sounded like the test data was part of the configuration, which is not true.

This can be changed in several different ways.

One is to `git clone` the repository locally and bind the repository folder to the `/app` folder in the container:

```shell
# volume will bind the local version of `optimade-python-tools` to the container.
docker run \
    --rm \
    --detach \
    --publish 8080:5000 \
    --env MAIN=main \
    --name my-optimade \
    --volume /path/to/optimade-python-tools:/app \
    ghcr.io/materials-consortia/optimade:latest
```

> Where you should change the `/path/to/optimade-python-tools` to the full local path to the repository - or use `$PWD` if you are running this command from the root of the cloned repository on a Unix system.

Reviewer: ```suggestion
You should change the `/path/to/optimade-python-tools` to the full local path to the repository - or use `$PWD` if you are running this command from the root of the cloned repository on a Unix system.
```

E.g., if you do not wish to use the test data you can add the `insert_test_data` key with a value of `false`.

Another option is to _not_ `git clone` the repository locally, but instead only generate a JSON or YAML file somewhere, where the directory of its location is bound to another location in the container that is not used, e.g., `/config` or similar.
As an example, let us say we generate `config.yml` locally, with the full path `/home/user/optimade/config.yml`.
Then we can start the server with the following command:

```shell
# volume will bind the local directory to an unused folder in the container.
docker run \
    --rm \
    --detach \
    --publish 8080:5000 \
    --env MAIN=main \
    --name my-optimade \
    --volume /home/user/optimade:/config \
    --env OPTIMADE_CONFIG_FILE=/config/config.yml \
    ghcr.io/materials-consortia/optimade:latest
```

As shown, it is necessary to update the environment variable `OPTIMADE_CONFIG_FILE` within the container to point to the new internal path to the config file.
By default, this environment variable points to `/app/optimade_config.json`.

> This also reveals another way of configuring the server: Use environment variables when running the container that supplies the values that would otherwise be supplied from the configuration file.

Reviewer: ```suggestion
This also reveals another way of configuring the server: set environment variables when running the container to supply values that would otherwise be supplied from the configuration file.
```

The `docker run` command even has the opportunity to pass a path to a file containing a list of environment variables (`--env-file /path/to/env_file`), if you wish to configure the server in this way instead of through the standard configuration file.

> ### **Example**: Multiple container services in closed network

Reviewer: This is useful stuff but the install docs are now very long. Could this (and perhaps some subset of the above notes) be moved to a `Deployment` section of the docs?

Author: Yeah, I was thinking about that as well. Just didn't have the overview to do it in the first run through. I'll try to update it, moving it around, when I get time.

Author: Done.

```

The following command starts a local MongoDB instance, runs the test suite, then stops and deletes the containers:

```shell
docker run -d --name mongo_test -p 27017:27017 -d mongo:4.4.6 \
&& OPTIMADE_DATABASE_BACKEND="mongodb" py.test; \
docker container stop mongo_test; docker container rm mongo_test
```

## Container image

### Retrieve the image

The [`optimade` container image](https://github.com/Materials-Consortia/optimade-python-tools/pkgs/container/optimade) is available from the [GitHub Container registry](https://ghcr.io).
To pull the latest version using [Docker](https://docs.docker.com/) run the following:

```shell

> docker pull ghcr.io/materials-consortia/optimade:latest

Reviewer 2: When I try to execute this line (nr. 126) with SUDO, I get the error: 

Error response from daemon: Head "https://ghcr.io/v2/materials-consortia/optimade/manifests/latest": unauthorized

I think the package is currently still set to private and we need to make it public.

Author: Yes, that's correct. This will happen when this PR is merged. See also [my comment above](https://github.com/Materials-Consortia/optimade-python-tools/pull/1171#issuecomment-1126167462).

Author: For now you can test it by logging in with a password that is a PAT (personal access token) with read permissions to packages :) See more [here](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry#authenticating-to-the-container-registry)
